pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	err = c.pruneRevisionHistory(operatorStatus)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getPrunerPodName(podName, nodeName string) string {
+	return fmt.Sprintf("prune-%s-%s", podName, nodeName)
+}
+
+func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.StaticPodOperatorStatus) error {
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		// Read status config map to figure which deployments to keep
+		statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(getStatusConfigMapName(nodeStatus.TargetRevision, nodeStatus.NodeName), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// store IDs as strings to join for protectedIDs param to pruner pod
+		var succeededRevisionIDs []string
+		var failedRevisionIDs []string
+
+		for revisionID, revisionStatus := range statusConfigMap.Data {
+			switch revisionStatus {
+			case string(corev1.PodSucceeded):
+				succeededRevisionIDs = append(succeededRevisionIDs, revisionID)
+			case string(corev1.PodFailed):
+				failedRevisionIDs = append(failedRevisionIDs, revisionID)
+			}
+		}
+		sort.Strings(succeededRevisionIDs)
+		sort.Strings(failedRevisionIDs)
+
+		err = c.pruneEligibleIDs(succeededRevisionIDs, failedRevisionIDs, c.succeededRevisionLimit, "succeeded", nodeStatus.NodeName)
+		if err != nil {
+			return err
+		}
+		err = c.pruneEligibleIDs(failedRevisionIDs, succeededRevisionIDs, c.failedRevisionLimit, "failed", nodeStatus.NodeName)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *PruneController) pruneEligibleIDs(revisionIDs, protectedRevisionIDs []string, revisionLimit int, podName, nodeName string) error {
+	// only prune if enough failed/succeeded revisions to prune
+	if len(revisionIDs) > revisionLimit {
+		// ids should be sorted, so looking back `revisionLimit` ids from the last element should give max id
+		maxEligibleRevision := revisionIDs[len(revisionIDs)-1-revisionLimit]
+		if err := c.ensurePrunePod(podName, nodeName, maxEligibleRevision, protectedRevisionIDs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *PruneController) ensurePrunePod(podName, nodeName, maxEligibleRevision string, protectedRevisions []string) error {
+	required := resourceread.ReadPodV1OrDie([]byte(prunerPod))
+	required.Name = getPrunerPodName(podName, nodeName)
+	required.Namespace = c.targetNamespace
+	required.Spec.NodeName = nodeName
+	required.Spec.Containers[0].Image = c.prunerPodImageFn()
+	required.Spec.Containers[0].Command = c.command
+	required.Spec.Containers[0].Args = append(required.Spec.Containers[0].Args,
+		fmt.Sprintf("-v=%d", 4),
+		fmt.Sprintf("--max-eligible-id=%s", maxEligibleRevision),
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
+		fmt.Sprintf("--resource-dir=%s", "/etc/kubernetes/static-pod-resources"),
+		fmt.Sprintf("--pod-manifest-dir=%s", "/etc/kubernetes/manifests"),
+		fmt.Sprintf("--static-pod-name=%s", c.staticPodName),
```

Comment:
also looks odd

Reply:
@deads2k similar question then, if this is being used by multiple operators this was just how I thought we could pass the name in